### PR TITLE
Impaler Burrower Strain

### PR DESCRIPTION
**Pull request by Shad0vvs**
## About The Pull Request

https://streamable.com/4m3rbr (some things are changed since this video)

A burrower strain that can deal damage while burrowed, causing enemies to have to be creative when attacking, removing its weeds will force it to surface.

### Impaler abilities:
**Burrowed Spikes:**
Description: Sends out fast spikes in a direction that hits tile by tile.
Cost: 50 Plasma
Cooldown: 8 Seconds.
Range: 5 Tiles
Damage: 35
**Can be used while burrowed.**

**Sunken Tail:**
Description: After a windup, strikes the targeted area damaging anyone in the tile.
Cost: 100 Plasma
Cooldown: 13 seconds.
Range: 7
Damage: 60
Windup: .8 Seconds

**Ensconce:**
Description: After a windup, go halfway into the ground in a fortified stance (prevents you from moving) which increases your vision range by 3, increases your Burrowed spikes ability range by 2, increases your Sunken Tail ability range by 2, and damage on both abilities by 30.
Modifies Sunken Tail to do an AoE effect:
https://streamable.com/xy0tq1 (middle is full damage + ensconce'ed damage bonus, edges are normal damage)

Cost: None
Cooldown: 3 Seconds
Buffs: Armor deflection increased by 25 and armor explosive resistance increased by 50 and small explosives cannot stun the target while in the stance.

**Base Tail Stab:**
**Can be used while burrowed.**

## Why It's Good For The Game

Xeno content to tric-I mean to help people have fun when playing Xenos.

## Changelog

:cl: Shad0vvs, Kugamo
add: Added new strain for the Burrower, "Impaler".
fix: Fixed burrowed mobs being able to be targeted by sentries, mines and SG autofire.
fix: Fixed burrowed mobs being able to grab mobs on the surface.
fix: Fixed burrowed mobs being able to resist while burrowed.
fix: Fixed burrowers taking damage from direct flame and shrapnel from explosions.
fix: Fixed burrowers being able to get slashed from enemy Xenos on the surface.
fix: Fixed burrowers unburrow stun to now properly target and stun enemy Xenos.
soundadd: Added sounds for base and Impaler burrower for burrowing and unburrowing, added sounds for Impaler burrower's attacks.
imageadd: Added sprites and effects for Impaler burrower by Kugamo.
/:cl:

### CODE REVIEW

**Comment by Shad0vvs** [comment]
Reminder: Need xenoid sprites before merge.

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 1]
```diff
@@ -24,6 +24,41 @@
...
+		RegisterSignal(src, COMSIG_LIVING_PREIGNITION, .proc/fire_immune)
+		RegisterSignal(src, list(
+			COMSIG_LIVING_FLAMER_CROSSED,
+			COMSIG_LIVING_FLAMER_FLAMED,
+		), .proc/flamer_crossed_immune)
+		ADD_TRAIT(src, TRAIT_ABILITY_BURROWED, TRAIT_SOURCE_ABILITY("Burrow"))
+		mouse_opacity = FALSE
+		playsound(src.loc, 'sound/effects/burrowing_s.ogg', 25)
+		update_icons()
+		addtimer(CALLBACK(src, .proc/do_burrow_cooldown), (caste ? caste.burrow_cooldown : 5 SECONDS))
```
Comment (fira) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:58]: wouldn't they always have a caste ?

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 2]
```diff
@@ -209,3 +271,270 @@
...
+		return
+
+	if(!check_clear_path_to_target(X, A, FALSE))
+		to_chat(X, SPAN_XENOWARNING("Something is blocking our path to [A]!"))
+		return
+
+	if(!check_plasma_owner())
+		return
+
+	var/turf/target = locate(A.x, A.y, A.z)
```
Comment (fira) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:399]: get_turf

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 3]
```diff
@@ -209,3 +271,270 @@
...
+		to_chat(xeno, SPAN_XENOWARNING("You can't do this while burrowed!"))
+		return
+
+	var/turf/T = get_turf(xeno)
+	var/obj/effect/alien/weeds/weeds = locate() in T
+	if(!weeds || !xeno.ally_of_hivenumber(weeds.hivenumber))
+		to_chat(xeno, SPAN_XENOWARNING("You need to do this on weeds!"))
+		return
+
+	if(!check_plasma_owner())
```
Comment (fira) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:453]: isn't plasma cost zero?

if you want to support changing it in the future, might want to also call check_and_use_plasma_owner

**Review by fira — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm (hunk 3 above)
Comment (Shad0vvs) [reply code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:486]: I want to change plasma after the do_after, just incase someone wants to put a plasma cost on it via ability.
Done via line 460

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm (hunk 1 above)
Comment (Shad0vvs) [reply code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:58]: I blame the burrower coder. will WIP.

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 4]
```diff
@@ -24,6 +24,41 @@
 
 	used_burrow = TRUE
 
+	if(mutation_type == BURROWER_IMPALER) //Impaler Burrower specific
```
[the remark this answers is not in the document]
Comment (Shad0vvs) [reply code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:27]: will wip

**Comment by Shad0vvs** [comment]
Made strain specific burrower have onclick burrow ability, modified Ensconce to onclick.
Added xenoid sprites.

Number changes:

Burrowed spikes:
CD from 10 > 8
Damage from 25 > 30
Ensconce'ed range bonus 1 > 2
Ensconce'ed damage bonus 10 > 30

Sunken tail:
CD from 15 > 13
Damage from 40 > 60
Windup delay from 1 Seconds > .8 seconds
Range from 5 > 7

Ensconce'ed range bonus 1 > 2
Ensconce'ed damage bonus 10 > 30
Added ability modification when Ensconce'ed:
https://streamable.com/xy0tq1 (middle is full damage + ensconce'ed damage bonus, edges are normal damage)

Ensconce:
CD from 5 > 3
Vision bonus from 9 > 10

**Comment by Thegreatspidernonbinary** [comment]
This is way too fast and gives nearly zero time to react

**Comment by Shad0vvs** [comment]
Burrowed Spikes:
Damage 30 > 35

Burrowed checks will happen every 0.5 seconds rather than every 1 seconds to give less leeway if you walk off weeds.

Explosions will affect the burrower now when burrowed, just not with damage. (The stun will be less than normal as they are underground)

**Comment by Shad0vvs** [comment]
Update test: Direct sprite clicks will damage the Xeno, otherwise the bullets will fly past them.

File: code/game/objects/effects/aliens.dm [hunk 5]
```diff
@@ -455,6 +455,18 @@
...
 	icon_state = "xeno_telegraph_lash"
 
+/obj/effect/xenomorph/ground_spike
+	name = "???"
+	desc = ""
+	icon_state = "ground_spike"
+	mouse_opacity = FALSE
+	var/mob/living/carbon/Xenomorph/source_xeno = null
+
+/obj/effect/xenomorph/ground_spike/New(loc, source_xeno = null, ttl = 10)
```
Comment (fira) [inline code/game/objects/effects/aliens.dm:465]: use Initialize

**Comment by Shad0vvs** [comment]
Removed Ensconce being a bodyblock but removed the ability to stack with it.

**Comment by morrowwolf** [comment]
Closing as per community ban.

**Pull request closed without merging**